source/date_time.c

```diff
@@ -564,7 +339,186 @@ static int s_parse_rfc_822(
+        return false;
+    }
+
+    aws_byte_cursor_advance(str, 1);
+    return true;
+}
+
+/* Read the (optional) fractional seconds (".123"). If present, str is advanced.
+ * Returns false if there was an error */
+static bool s_read_optional_fractional_seconds(struct aws_byte_cursor *str) {
+    if (str->len == 0) {
+        return true;
+    }
+
+    uint8_t c = str->ptr[0];
+    if (c != '.' && c != ',') {
+        return true;
+    }
+
+    size_t num_digits = 0;
+    for (size_t i = 1; i < str->len; ++i) {
+        if (aws_isdigit(str->ptr[i])) {
+            ++num_digits;
+        } else {
+            break;
+        }
+    }
+
+    if (num_digits == 0) {
+        return false;
+    }
+
+    aws_byte_cursor_advance(str, 1 + num_digits);
+    return true;
+}
+
+/* Parses ISO 8601, both extended and basic format are accepted.
+ * Returns true if successful. */
+static bool s_parse_iso_8601(struct aws_byte_cursor str, struct tm *parsed_time, time_t *seconds_offset) {
+    AWS_ZERO_STRUCT(*parsed_time);
+    *seconds_offset = 0;
+    uint8_t c = 0;
+
+    /* read year */
+    if (!s_read_n_digits(&str, 4, &parsed_time->tm_year)) {
+        return false;
+    }
+    parsed_time->tm_year -= 1900;
+
+    /* be lenient, allow date with separator or not */
+    bool has_date_separator = s_advance_if_c(&str, '-');
+
+    /* read month */
+    if (!s_read_n_digits(&str, 2, &parsed_time->tm_mon)) {
+        return false;
+    }
+    parsed_time->tm_mon -= 1;
+
+    if (has_date_separator) {
+        if (!s_read_1_char(&str, &c) || c != '-') {
+            return false;
+        }
+    }
+
+    /* read month-day */
+    if (!s_read_n_digits(&str, 2, &parsed_time->tm_mday)) {
+        return false;
+    }
+
+    /* ISO8601 supports date only with no time portion */
+    if (str.len == 0) {
+        return true;
+    }
+
+    /* followed by T or space (allowed by rfc3339#section-5.6) */
+    if (!s_read_1_char(&str, &c) || !(tolower(c) == 't' || c == ' ')) {
+        return false;
+    }
+
+    /* read hours */
+    if (!s_read_n_digits(&str, 2, &parsed_time->tm_hour)) {
+        return false;
+    }
+
+    /* be lenient, allow time with separator or not */
+    bool has_time_separator = s_advance_if_c(&str, ':');
+
+    /* read minutes */
+    if (!s_read_n_digits(&str, 2, &parsed_time->tm_min)) {
+        return false;
+    }
+
+    if (has_time_separator) {
+        if (!s_read_1_char(&str, &c) || c != ':') {
+            return false;
+        }
+    }
+
+    /* read seconds */
+    if (!s_read_n_digits(&str, 2, &parsed_time->tm_sec)) {
+        return false;
+    }
+
+    /* fractional seconds are optional (discard value since tm struct has no corresponding field) */
+    if (!s_read_optional_fractional_seconds(&str)) {
+        return false;
+    }
+
+    /* read final Z, or (+/-) indicating there will be an offset */
+    if (!s_read_1_char(&str, &c)) {
+        return false;
+    }
+
+    if (tolower(c) == 'z') {
```

Comment:
Do we wanna verify that there is not anything extra at the end?

Reply:
Good thinking, I thought the same. But the old code didn't check it, so I didn't want to start checking it now. I don't want some trailing space in a header to start failing now when it used to pass